### PR TITLE
[docker] Rework incremental build

### PR DESCRIPTION
**Pull request by mgorven**
Restore original validator Docker config, and move incremental build to
`validator-dev`. Instead of having separate Dockerfiles with
duplication, use `awk` to transform validator.Dockerfile into the ones
needed for incremental build.

Test Plan: Built `validator` and `validator-dev`.

### CODE REVIEW

**Comment by mgorven** [comment]
@andll Could you test the incremental build on OS X please?

**Comment by andll** [comment]
(1) I think the way it is right now makes incremental compilation second class citizen - whoever updates docker file won't even know this feature exist and will break it, and only during outage oncall will find out that it does not work as expected.

(2) More importantly - how does build on Circle going to work? @sausagee currently have two docker files, are they in future going to be produced using awk too?

**Comment by andll** [comment]
To be honest more I think about it less I like this awk approach
Looks like CI needs separate files for caching and incremental build
And local incremental build needs separate files too

So what do we use this single docker file for at this point?

**Comment by mgorven** [comment]
(1) Given that incremental compilation goes against the core principles of Docker's design, it is inherently a second class citizen. While the `awk` processing is not completely robust, there's only a small subset of changes to the docker file which would break it, and the common changes (like adding cmdline args) will not. Given that an outage right now is just the testnet, I don't think we need to prioritise incident response time over doing this right.

(2) I don't want Circle to use different Docker files either, and I have concerns with the changes there as well. My next step after this diff was going to be to migrate to the Docker files in libra-ops.

> So what do we use this single docker file for at this point?

There are also external developers, and we should be providing a proper Docker file for them to use.

File: docker/validator-dev/build.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+set -e
+
+DIR="$( cd "$( dirname "$0" )" && pwd )"
```
Comment (huitseeker) [inline docker/validator-dev/build.sh]:
```suggestion
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
([rationale](https://stackoverflow.com/a/35006505/47978))

**Comment by sausagee** [comment]
I am sharing notes from the discussion with Andrey, Michael and Brett earlier today to give additional context.  We all want fast and reliable builds.  When we need to bisect through running cluster tests, we want to iterate many builds quickly.  We can build these images locally on a Mac or on Circle.  But in either case, the standard docker build flow is too slow for us.  For Circle, it has its own limitation on remote docker container size, which takes 1 hr to build. That was the primary motivation behind the incremental build flow and the build flow where I stitched the dockage images.  We understand these build flows are not officially supported and their build artifacts will never see the light of production. They are only used in bisecting and testing, and we will tag them as such. 

I do feel the need to have a single source of truth for the dockerfile(s) though, considering the port configs, additional py packages we need.  Keeping multiple copies in `validator` and `validator-dev` seems error prone, at least human errors. Let's say someone needs to update the port config, how does the person supposed to know/remember to keep all copies consistent.

It seems to me there are three distinct stages in the docker build flow: 
1. bootstrap the base image and install the toolchain
2. build rust package
3. config final image
It looks like the 2nd step is where these build flows differ and where we are trying to optimize. Docker has multi-staged build, but they have to be in one single docker file.  Docker doesn't have feature to include another docker file, but that seems what we need here.

@mgorven and @andll, can we consider splitting these stages into its own docker file and using a build script to generate the final docker file to be used in 1) prod build and 2) test build?

**Comment by mgorven** [comment]
@sausagee This doesn't have separate copies in `validator-dev`, it transforms the one in `validator` by splitting out those three stages (using the various `FROM` lines now). If someone changes the ports/cmdline/args everything will keep working.

I'd really like to get a proper Dockerfile back into the libra repo, we can make further improvements, but in the meantime this gets us back to a sane state.

**Comment by andll** [comment]
I think one thing we can consider is 
- having separate docker files
- script that stitches them together (stitching is much less error prone then split with awk)
- Keep both separate files and stitched file in repo and have CI check that they are matching (basically, re-run stitching script on CI and check that `git diff` is empty)

**Comment by opsguy** [comment]
We really need to focus on the primary and most important use case here: Making it as easy and reliable as possible for anyone to build/deploy this project.

Speed is really a secondary concern here, and if we are ever trading off reliability and readability for speed I think we are doing the project a disservice. Especially if the primary driver for the speed concern is single CI tool. I fully understand how important and critical speed can be for developer efficiency, and am actively working on other solutions for this. 

With keeping the primary focus on reliability, I think we are really obligated to return to a single readable Dockerfile until we can figure something else out. Given how long we were in the previous state and how new this multi-stage change is, I think granting a stay here and landing this is the most prudent course of action. Especially since we had previously discussed this at length and decided against it. 

Barring any universal agreement on keeping the new changes, I will go ahead and accept this by EOD Friday. I believe in general that any controversial changes should be agreed upon before landing, and when such situations arise I think the standard process should be going to the previous state until such universal agreement is reached rather than leaving the contested changes live. If anyone disagrees with this as a general principle please elaborate here and we can decide how to proceed on these in the future.

**Comment by huitseeker** [comment]
With more context, the single multi-staged Dockerfile proposed here makes sense: it embodies doing w/ a Dockerfile all that can be done with that — and since that file serves a bigger role than dev tests, it seems good to have it as a baseline.

Granted it's also important to achieve good efficiency besides (on incrementality or CI build speed), but I think we can iterate on top of this PR using non-Dockerfile means to achieve those goals, including e.g. :
- [sccache](https://github.com/mozilla/sccache),
- building with [buildah](https://github.com/containers/buildah/blob/master/docs/buildah-bud.md#description), which supports includes,

**Comment by andll** [comment]
I just want to remind everyone where did this incremental compilation diff came from:
- Last week @dmitri-perelman spent weekend figuring out what was going on
- Previously I spent 3 days fighting similar issue with consensus deadlock, because every build takes a lot of time and there are a lot of hypothesis to check

I don't really see point in merging this diff to be honest, we can as well just revert incremental compilation diff completely and hope we have better solution by the time we have next outage

**Comment by opsguy** [comment]
Appreciate everyone's understanding here. I think we're close on getting faster builds working anyways so hopefully this is moot in the near future.

**Pull request closed without merging**